core/GlobalState.h

```diff
@@ -320,6 +321,10 @@ class GlobalState final {
 
     bool shouldReportErrorOn(Loc loc, ErrorClass what) const;
 
+    void clearErrorCacheForFile(core::FileRef fref,
+                                std::function<bool(const std::unique_ptr<core::ErrorQueueMessage> &)> predicate);
+    UnorderedMap<core::FileRef, std::vector<std::unique_ptr<core::ErrorQueueMessage>>> errors;
```

Comment:
This looks somewhat alarming.

- I don't think this name is accurate?

  If I'm reading the code right, it's more like a cache of errors for files we've flushed once already but need to hold onto for later just in case.
  
- Regardless, we should not be exposing a public, read/write, mutable field which is error-related on GlobalState. Errors are the one place in `GlobalState` where we "violate" const correctness with the `mutable` keyword: 
https://github.com/sorbet/sorbet/blob/ce3d1ad38d8ef83615b8015848bec465c0fb9f2c/core/GlobalState.h#L255

  This means that even if you have `const GlobalState &`, you're still allowed to mutate `GlobalState`, so long as you only mutate this field. This is okay because `ErrorQueue` is explicitly designed to work in a multi-threaded context. Specifically, all error storage inside an ErrorQueue is mediated by `ConcurrentUnBoundedQueue`. The exception is the `collected` field, which is guarded in all cases by calls to `checkOwned` to make sure that the mutable, unguarded access to the ErrorQueue only happens from the main thread (i.e., the thread that created that ErrorQueue).
  
  Note that `collected` itself is owned by `ErrorQueue`, and is private, thus hidden from people interacting with GlobalState (the opposite of what `GlobalState::errors` is right now: a publicly exposed, possibly mutable point.

I think this suggests that we want to be using ErrorQueue to both store this cache and make sure that the access is private and guarded by checkOwned.

Reply:
I see now that this is the reason why the `ErrorQueueMessage` needs a clone method, because you're running into problems with the fact that making `GlobalState` own these errors means that you can't steal them in order to flush them, because typecheck operates on a `const GlobalState &`.

I'm pretty sure that if ownership of these cached errors were to move to ErrorQueue, it would be possible to steal the errors instead of needing to clone them.

Reply:
> I don't think this name is accurate?
> If I'm reading the code right, it's more like a cache of errors for files we've flushed once already but need to hold onto for later just in case.

Would LSPErrorCache make more sense?

Reply:
>I'm pretty sure that if ownership of these cached errors were to move to ErrorQueue, it would be possible to steal the errors instead of needing to clone them.

That was my first approach. It's hard to make it work properly, because we swap out `errorQueue` a bunch of times in Sorbet.
For example on a fast path:
https://github.com/sorbet/sorbet/blob/a9744c21e4e2affb12f2d60b18f9654d375b1e84/main/lsp/LSPTypechecker.cc#L217-L218

And slow path:
https://github.com/sorbet/sorbet/blob/a9744c21e4e2affb12f2d60b18f9654d375b1e84/main/lsp/LSPTypechecker.cc#L427-L429

And in the LSPQuery mechanics too:
https://github.com/sorbet/sorbet/blob/a9744c21e4e2affb12f2d60b18f9654d375b1e84/main/lsp/LSPTypechecker.cc#L681-L683

Storing the cache in the errorQueue led to loss of errors because of those swaps

Reply:
>I see now that this is the reason why the ErrorQueueMessage needs a clone method, because you're running into problems with the fact that making GlobalState own these errors means that you can't steal them in order to flush them, because typecheck operates on a const GlobalState &.

That's the gist of the issue, yeah. I also probably can't steal them per se.  Here's why:
Publishing a diagnostic requires a transfer of ownership of ErrorQueueMessage, because errorFlusher takes a vector of unique_ptrs
https://github.com/sorbet/sorbet/blob/a9744c21e4e2affb12f2d60b18f9654d375b1e84/core/ErrorQueue.cc#L77

In the duration of the whole typecheking pipeline the namer error should be reported three times: after namer, resolver and typecheck. So that means I can't flush the same ErrorQueueMessage instance three times. If I could steal the error it would be freed after `errorFlusher->flushErrors`, am I getting this right? In that case I need a copy to keep flushing the error